File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlaylistPlayQueue.java
package org.schabi.newpipe.player.playqueue;

import org.schabi.newpipe.extractor.Page;
import org.schabi.newpipe.extractor.playlist.PlaylistInfo;
import org.schabi.newpipe.extractor.playlist.PlaylistInfoItem;
import org.schabi.newpipe.extractor.stream.StreamInfoItem;
import org.schabi.newpipe.util.ExtractorHelper;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public final class PlaylistPlayQueue extends AbstractInfoPlayQueue<PlaylistInfo, PlaylistInfoItem> {
    public PlaylistPlayQueue(final PlaylistInfoItem item) {
        super(item);
    }

    public PlaylistPlayQueue(final PlaylistInfo info) {
        this(info.getServiceId(), info.getUrl(), info.getNextPage(), info.getRelatedItems(), 0);
    }

    public PlaylistPlayQueue(final int serviceId,
                             final String url,
                             final Page nextPage,
                             final List<StreamInfoItem> streams,
                             final int index) {
        super(serviceId, url, nextPage, streams, index);
    }

    @Override
    protected String getTag() {
        return "PlaylistPlayQueue@" + Integer.toHexString(hashCode());
    }

    @Override
    public void fetch() {
        if (this.isInitial) {
            ExtractorHelper.getPlaylistInfo(this.serviceId, this.baseUrl, false)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(getHeadListObserver());
        } else {
            ExtractorHelper.getMorePlaylistItems(this.serviceId, this.baseUrl, this.nextPage)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(getNextPageObserver());
        }
    }
}

File: app/src/main/java/org/schabi/newpipe/player/playqueue/PlaylistPlayQueue.kt
package org.schabi.newpipe.player.playqueue

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.schabi.newpipe.extractor.Page
import org.schabi.newpipe.extractor.playlist.PlaylistInfo
import org.schabi.newpipe.extractor.stream.StreamInfoItem
import org.schabi.newpipe.util.ExtractorHelper

class PlaylistPlayQueue : AbstractInfoPlayQueue<PlaylistInfo> {

    constructor(info: PlaylistInfo) : super(info)
    constructor(info: PlaylistInfo, index: Int) : super(info, index)
    constructor(
        serviceId: Int,
        url: String,
        nextPage: Page?,
        streams: List<StreamInfoItem>,
        index: Int
    ) : super(serviceId, url, nextPage, streams, index)

    override fun getTag(): String = "PlaylistPlayQueue@${Integer.toHexString(hashCode())}"

    override fun fetch() {
        startFetch {
            try {
                if (isInitial) {
                    val result = withContext(Dispatchers.IO) {
                        ExtractorHelper.getPlaylistInfo(serviceId, baseUrl, false)
                    }
                    handleResult(result)
                } else {
                    val result = withContext(Dispatchers.IO) {
                        ExtractorHelper.getMorePlaylistItems(serviceId, baseUrl, nextPage!!)
                    }
                    handleNextPageResult(result)
                }
            } catch (e: Throwable) {
                handleError(e)
            }
        }
    }
}
